File: packages/bot/src/commands/recent.ts
import type { Context } from "grammy";
import { getUser, listRecentExpenses } from "@nudge/db";
import type { Db } from "@nudge/db";
import { formatAmount } from "../utils/format";

export function recentCommand(db: Db) {
  return async (ctx: Context) => {
    const user = await getUser(db, BigInt(ctx.from!.id));
    if (!user) return;

    const items = await listRecentExpenses(db, user.id, 10);
    if (items.length === 0) {
      await ctx.reply("No expenses logged yet.");
      return;
    }

    const lines = items.map((e, i) => {
      const cat = e.category;
      const emoji = cat?.emoji ?? "📦";
      const name = cat?.name ?? "Misc";
      const date = e.spentAt.toLocaleDateString("en-IN", {
        day: "numeric",
        month: "short",
      });
      return `${i + 1}. ${emoji} ${formatAmount(e.amount, e.currency)} → ${name}  · ${date}${e.merchant ? `  (${e.merchant})` : ""}`;
    });

    await ctx.reply(`Recent expenses:\n\n${lines.join("\n")}`);
  };
}

File: apps/api/bot/commands/recent.py
from telegram import Update
from telegram.ext import ContextTypes

from db.queries import get_user, list_recent_expenses
from bot.utils.format import format_amount


async def recent_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    pool = context.bot_data["pool"]
    user = await get_user(pool, update.effective_user.id)
    if not user:
        return

    items = await list_recent_expenses(pool, str(user["id"]), 10)
    if not items:
        await update.message.reply_text("No expenses logged yet.")
        return

    lines = []
    for i, e in enumerate(items, 1):
        emoji = e.get("category_emoji") or "📦"
        name = e.get("category_name") or "Misc"
        date = e["spent_at"].strftime("%-d %b")
        merchant = f"  ({e['merchant']})" if e.get("merchant") else ""
        lines.append(
            f"{i}. {emoji} {format_amount(e['amount'], e['currency'])} → {name}  · {date}{merchant}"
        )

    await update.message.reply_text("Recent expenses:\n\n" + "\n".join(lines))
